File: akka-docs/src/test/java/jdocs/stream/operators/MergeSequenceDocExample.java
/*
 * Copyright (C) 2020-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package jdocs.stream.operators;

import akka.actor.ActorSystem;

// #import
import akka.NotUsed;
import akka.japi.Pair;
import akka.stream.ClosedShape;
import akka.stream.UniformFanInShape;
import akka.stream.UniformFanOutShape;
import akka.stream.javadsl.Flow;
import akka.stream.javadsl.GraphDSL;
import akka.stream.javadsl.MergeSequence;
import akka.stream.javadsl.Partition;
import akka.stream.javadsl.RunnableGraph;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;
// #import

public class MergeSequenceDocExample {

  private final ActorSystem system = ActorSystem.create("MergeSequenceDocExample");

  interface Message {}

  boolean shouldProcess(Message message) {
    return true;
  }

  Source<Message, NotUsed> createSubscription() {
    return null;
  }

  Flow<Pair<Message, Long>, Pair<Message, Long>, NotUsed> createMessageProcessor() {
    return null;
  }

  Sink<Message, NotUsed> createMessageAcknowledger() {
    return null;
  }

  void mergeSequenceExample() {

    // #merge-sequence

    Source<Message, NotUsed> subscription = createSubscription();
    Flow<Pair<Message, Long>, Pair<Message, Long>, NotUsed> messageProcessor =
        createMessageProcessor();
    Sink<Message, NotUsed> messageAcknowledger = createMessageAcknowledger();

    RunnableGraph.fromGraph(
            GraphDSL.create(
                builder -> {
                  // Partitions stream into messages that should or should not be processed
                  UniformFanOutShape<Pair<Message, Long>, Pair<Message, Long>> partition =
                      builder.add(
                          Partition.create(2, element -> shouldProcess(element.first()) ? 0 : 1));
                  // Merges stream by the index produced by zipWithIndex
                  UniformFanInShape<Pair<Message, Long>, Pair<Message, Long>> merge =
                      builder.add(MergeSequence.create(2, Pair::second));

                  builder.from(builder.add(subscription.zipWithIndex())).viaFanOut(partition);
                  // First goes through message processor
                  builder.from(partition.out(0)).via(builder.add(messageProcessor)).viaFanIn(merge);
                  // Second partition bypasses message processor
                  builder.from(partition.out(1)).viaFanIn(merge);

                  // Unwrap message index pairs and send to acknowledger
                  builder
                      .from(merge.out())
                      .to(
                          builder.add(
                              Flow.<Pair<Message, Long>>create()
                                  .map(Pair::first)
                                  .to(messageAcknowledger)));

                  return ClosedShape.getInstance();
                }))
        .run(system);

    // #merge-sequence
  }
}

File: akka-docs/src/test/scala/docs/stream/operators/MergeSequenceDocExample.scala
/*
 * Copyright (C) 2020-2025 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.stream.operators

import akka.actor.ActorSystem

object MergeSequenceDocExample {

  implicit val system: ActorSystem = ???

  // #merge-sequence
  import akka.NotUsed
  import akka.stream.ClosedShape
  import akka.stream.scaladsl.{ Flow, GraphDSL, MergeSequence, Partition, RunnableGraph, Sink, Source }

  val subscription: Source[Message, NotUsed] = createSubscription()
  val messageProcessor: Flow[(Message, Long), (Message, Long), NotUsed] =
    createMessageProcessor()
  val messageAcknowledger: Sink[Message, NotUsed] = createMessageAcknowledger()

  RunnableGraph
    .fromGraph(GraphDSL.create() { implicit builder =>
      import GraphDSL.Implicits._
      // Partitions stream into messages that should or should not be processed
      val partition = builder.add(Partition[(Message, Long)](2, {
        case (message, _) if shouldProcess(message) => 0
        case _                                      => 1
      }))
      // Merges stream by the index produced by zipWithIndex
      val merge = builder.add(MergeSequence[(Message, Long)](2)(_._2))

      subscription.zipWithIndex ~> partition.in
      // First goes through message processor
      partition.out(0) ~> messageProcessor ~> merge
      // Second partition bypasses message processor
      partition.out(1) ~> merge
      merge.out.map(_._1) ~> messageAcknowledger
      ClosedShape
    })
    .run()

  // #merge-sequence

  def shouldProcess(message: Message): Boolean = true
  trait Message
  def createSubscription(): Source[Message, NotUsed] = ???
  def createMessageProcessor(): Flow[(Message, Long), (Message, Long), NotUsed] = ???
  def createMessageAcknowledger(): Sink[Message, NotUsed] = ???

}
